### DISCUSSION TITLE
有线表格恢复的算法是不是有问题啊？怎么改善？

### DISCUSSION THREAD

**Discussion by nissansz** [RapidAI/TableStructureRec#6] [Q&A]
有线表格恢复的算法是不是有问题啊？怎么改善？
![21](https://github.com/RapidAI/TableStructureRec/assets/41010669/7bc1203b-2229-42e2-9eca-18177cf16fb0)
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/5ffca865-ff5c-4a62-a633-85dbc977cddb)
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/0674066f-b859-4e75-90d0-fe74f9eee211)

**Comment by qqeip** [comment]
这种还是先cls转好角度再处理

**Reply by nissansz** [reply]
原图是这样的，识别前已经校正倾斜了。表格恢复的算法能改吗？怎么改？
![12](https://github.com/RapidAI/TableStructureRec/assets/41010669/44dc02b3-7e45-4379-8dd1-38bc8d304467)

**Comment by SWHL** [comment]
这个是后处理的问题，需要去看后处理源码，排查。

**Reply by nissansz** [reply]
有的表有时会在角落残缺，识别不到表格线，不知道怎么推理出正确表格结构。
PaddleStructure模型好像可以自动加上
![21](https://github.com/RapidAI/TableStructureRec/assets/41010669/ee15320b-000b-4918-ab4c-8d76d76f273a)

**Comment by qqeip** [comment]
你可以直接用rapid_table 试试
<html><body><table border="1"><tr><td colspan="2">合单单号：BD23050100013 客户名种：上海松江万达店</td><td colspan="7">一铭食品（上海）销售单 合单时间：2023-05-0121:3337 南地地址：上海市松江区广鼠林路658弄172、173号（万 送产场金街） 联系人：许系验 手机号码：13471660339路线：</td></tr><tr><td colspan="2">订单新注：</td><td></td><td></td><td></td><td></td><td></td><td></td><td></td></tr><tr><td>序号 百</td><td>商品编码</td><td>品牌</td><td>商品名称</td><td>商品规格</td><td>数量</td><td>单位</td><td>单价</td><td>州</td></tr><tr><td></td><td>2000109</td><td>安井</td><td>2.5kg*4散装水晶包</td><td>2.5kg</td><td>1</td><td>袋</td><td>33</td><td>33</td></tr><tr><td>2</td><td>2000115</td><td>安井</td><td>2.5kg*4散装脆骨啵破肠</td><td>2.5kg</td><td>1</td><td>袋</td><td>43.5</td><td>43.5</td></tr><tr><td>3</td><td>2000243</td><td>蓝新</td><td>上汤百叶包</td><td>2.5kg</td><td>$</td><td>包</td><td>38</td><td>38</td></tr><tr><td>4</td><td>2000229</td><td>立圆</td><td>立圆北极翅</td><td>2.5kg</td><td>丰</td><td>包</td><td>65</td><td>65</td></tr><tr><td>5</td><td>2000198</td><td>升隆</td><td>鱼子福袋</td><td>2.5kg</td><td></td><td>袋</td><td>160</td><td>160</td></tr><tr><td>6</td><td>2000157</td><td>升隆</td><td>玉米香肠</td><td>1.5kg</td><td>2</td><td></td><td>23</td><td>46</td></tr><tr><td>1</td><td>2000319</td><td>源香</td><td>烤鱼棒</td><td>2.5kg</td><td>牛</td><td>包</td><td>45</td><td>45</td></tr><tr><td>8</td><td>2000251</td><td>源香</td><td>嘉庚亲亲肠</td><td>2.5kg 2.2kg*5</td><td>千 十</td><td>包 箱</td><td>40 48</td><td>48</td></tr><tr><td>9 10</td><td>2000282</td><td>浩野</td><td>魔芋丝结</td><td>3.75kg</td><td>十</td><td>箱</td><td>70</td><td>70</td></tr><tr><td>2000315</td><td></td><td>仁鸿</td><td>豆皮王中王 酿香大红肠</td><td>0.85kg</td><td>丰</td><td>根</td><td>28</td><td>28 23</td></tr></table></body></html>

**Reply by nissansz** [reply]
是的，rapidtable有时结果好，有时差，不知道怎么通过代码判断，哪个结果好，该用哪个方法。
RapidTable有时连很清晰的表格线都识别有问题
![table2](https://github.com/RapidAI/TableStructureRec/assets/41010669/04ef644f-f370-4731-a0df-4b2d216e26d2)
![table2](https://github.com/RapidAI/TableStructureRec/assets/41010669/5e81799e-9166-4b81-8a5a-d26688fe4781)

**Reply by nissansz** [reply]
有没有哪里有这些前处理的完整代码？
特别是文字行变形校正之类
https://medium.com/@Gekko_lab/make-your-ocr-results-more-accurate-part-ii-preprocessing-3d212ae16191

**Reply by nissansz** [reply]
还有就是，有没有什么办法修补表格线，让tablestructure能更好地识别出正确结构。

**Reply by qqeip** [reply]
pytorch里都有完整的代码 
https://learnopencv.com/trocr-getting-started-with-transformer-based-ocr/

**Comment by qqeip** [comment]
**# Augmentations.
train_transforms = transforms.Compose([
    transforms.ColorJitter(brightness=.5, hue=.3),
    transforms.GaussianBlur(kernel_size=(5, 9), sigma=(0.1, 5)),
])**

class CustomOCRDataset(Dataset):
    def __init__(self, root_dir, df, processor,dicts, max_target_length=128):
        self.root_dir = root_dir
        self.df = df
        self.processor = processor
        self.dicts = dicts
        self.max_target_length = max_target_length

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        # The image file name.
        file_name = self.df['file_name'][idx]
        # The text (label).
        text = self.df['text'][idx]
        # Read the image, apply augmentations, and get the transformed pixels.
        image = Image.open(self.root_dir + file_name).convert('RGB')
        **image = train_transforms(image)**
        pixel_values = self.processor(image, return_tensors='pt').pixel_values
        # Pass the text through the tokenizer and get the labels,
        # i.e. tokenized labels.
        labels = self.processor.tokenizer(
            text,
            padding='max_length',
            max_length=self.max_target_length
        ).input_ids
        # We are using -100 as the padding token.
        labels = [label if label != self.processor.tokenizer.pad_token_id else -100 for label in labels]
        encoding = {"pixel_values": pixel_values.squeeze(), "labels": torch.tensor(labels)}
        return encoding

**Reply by nissansz** [reply]
这是什么代码？
